src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+                .optional()
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+}
+
+impl ContractDatabase for SqliteContractDatabase {
+    fn create_map(&mut self, contract_name: &str, map_name: &str, key_type: TupleTypeSignature, value_type: TupleTypeSignature) -> Result<()> {
+        let key_type = TypeSignature::new_atom(AtomTypeIdentifier::TupleType(key_type));
+        let value_type = TypeSignature::new_atom(AtomTypeIdentifier::TupleType(value_type));
+
+        self.execute("INSERT INTO maps_table (contract_name, map_name, key_type, value_type) VALUES (?, ?, ?, ?)",
+                     &[contract_name, map_name, &key_type.serialize()?, &value_type.serialize()?])?;
+        Ok(())
+    }
+
+    fn fetch_entry(&self, contract_name: &str, map_name: &str, key: &Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), (*key).clone())))
+        }
+
+        let params: [&ToSql; 2] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?];
+
+        let sql_result: Option<Option<String>> = 
+            self.query_row(
+                "SELECT value FROM data_table WHERE map_identifier = ? AND key = ? ORDER BY data_identifier DESC",
+                &params,
+                |row| {
+                    row.get(0)
+                })?;
+        match sql_result {
+            None => {
+                Ok(Value::Void)
+            },
+            Some(sql_result) => {
+                match sql_result {
+                    None => Ok(Value::Void),
+                    Some(value_data) => Value::deserialize(&value_data)
+                }
+            }
+        }
+    }
+
+    fn set_entry(&mut self, contract_name: &str, map_name: &str, key: Value, value: Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(&key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), key)))
+        }
+        if !map_descriptor.value_type.admits(&value) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.value_type), value)))
+        }
+
+        let params: [&ToSql; 3] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?,
+                                   &Some(value.serialize()?)];
+
+        self.execute(
+            "INSERT INTO data_table (map_identifier, key, value) VALUES (?, ?, ?)",
+            &params)?;
+
+        return Ok(Value::Void)
+    }
+
+    fn insert_entry(&mut self, contract_name: &str, map_name: &str, key: Value, value: Value) -> Result<Value> {
```

Comment:
Since this is doing a test-and-set, should this operation be carried out within a Sqlite transaction?

Reply:
It could be done within a savepoint, since these actions are all already being done within savepoints (I don't think you can nest _transactions_ within savepoints, only more savepoints). But I'm pretty sure that it won't change any of the semantics of the test-and-set.

Reply:
One thing that could be changed, though, is we could have rusqlite actually enforce that all this is happening in savepoints, by using the savepoint rusqlite structs instead of manually managing our savepoints.

Reply:
Gotcha.  I think that would be more appropriate, if it's convenient to do so.  For example, in other parts of the code, if transaction-like semantics are expected, then the method takes an explicit `&mut Transaction<'a>` argument to ensure that all database operations occur within that transaction.  It might make more sense to do similar with Rusqlite savepoints.

Reply:
Yep, it should be pretty straight forward to do so.